README.md
# BaNaNaS API

[![GitHub License](https://img.shields.io/github/license/OpenTTD/bananas-api)](https://github.com/OpenTTD/bananas-api/blob/master/LICENSE)
[![GitHub Tag](https://img.shields.io/github/v/tag/OpenTTD/bananas-api?include_prereleases&label=stable)](https://github.com/OpenTTD/bananas-api/releases)
[![GitHub commits since latest release](https://img.shields.io/github/commits-since/OpenTTD/bananas-api/latest/master)](https://github.com/OpenTTD/bananas-api/commits/master)

[![GitHub Workflow Status (Testing)](https://img.shields.io/github/workflow/status/OpenTTD/bananas-api/Testing/master?label=master)](https://github.com/OpenTTD/bananas-api/actions?query=workflow%3ATesting)
[![GitHub Workflow Status (Publish Image)](https://img.shields.io/github/workflow/status/OpenTTD/bananas-api/Publish%20image?label=publish)](https://github.com/OpenTTD/bananas-api/actions?query=workflow%3A%22Publish+image%22)
[![GitHub Workflow Status (Deployments)](https://img.shields.io/github/workflow/status/OpenTTD/bananas-api/Deployment?label=deployment)](https://github.com/OpenTTD/bananas-api/actions?query=workflow%3A%22Deployment%22)

[![GitHub deployments (Staging)](https://img.shields.io/github/deployments/OpenTTD/bananas-api/staging?label=staging)](https://github.com/OpenTTD/bananas-api/deployments)
[![GitHub deployments (Production)](https://img.shields.io/github/deployments/OpenTTD/bananas-api/production?label=production)](https://github.com/OpenTTD/bananas-api/deployments)


This is the HTTP API for OpenTTD's content service, called BaNaNaS.
It works together with [bananas-server](https://github.com/OpenTTD/bananas-server), which serves the in-game client.

## Development

This API is written in Python 3.8 with aiohttp, and makes strong use of asyncio.

### Running a local server

#### Dependencies

- Python3.8 or higher.
- [tusd](https://github.com/tus/tusd). For example, copy the `tusd` binary in your `~/.local/bin`.

#### Preparing your venv

To start it, you are advised to first create a virtualenv:

```bash
python3 -m venv .env
.env/bin/pip install -r requirements.txt
```

#### Starting a local server

You can start the HTTP server by running:

```bash
.env/bin/python -m bananas_api --web-port 8080 --tusd-port 1080 --storage local --index local --user developer
```

This will start the API on port 8080 for you to work with locally.

### Running via docker

```bash
docker build -t openttd/bananas-api:local .
export BANANAS_COMMON=$(pwd)/../bananas-common
mkdir -p "${BANANAS_COMMON}/local_storage" "${BANANAS_COMMON}/BaNaNaS"
docker run --rm -p 127.0.0.1:8080:80 -p 127.0.0.1:1080:1080 -v "${BANANAS_COMMON}/local_storage:/code/local_storage" -v "${BANANAS_COMMON}/BaNaNaS:/code/BaNaNaS" openttd/bananas-api:local
```

The mount assumes that [bananas-server](https://github.com/OpenTTD/bananas-server) and this repository has the same parent folder on your disk, as both servers need to read the same local storage.

### Files upload (tusd)

tusd runs on its own port (1080 by default), and listens on `/new-package/files`.
With other words: the webserver does not forward that URL to tusd.
This means that for clients, you need to contact two endpoints:

- the web-port for everything except `/new-package/files`.
- the tusd-port for `/new-package/files`.

In production the Load Balancer redirects the URLs to the right ports, but during development this is something to keep in mind.

[bananas-frontend-cli](https://github.com/OpenTTD/bananas-frontend-cli) for example allows you to define the web-endpoint and the tusd-endpoint.


regression/changelog.txt
This is a changelog designed for the regression test.

We changed all the ✈ into 🚢 for the README.


bananas_api/__main__.py
import asyncio
import click
import ctypes
import logging
import signal

from aiohttp import web
from aiohttp.web_log import AccessLogger
from ctypes.util import find_library

from .helpers.click import click_additional_options
from .helpers.content_save import click_content_save
from .helpers.sentry import click_sentry
from .helpers.user_session import (
    click_user_session,
    register_webroutes,
)
from .new_upload.session import click_cleanup_graceperiod
from .new_upload.session_publish import click_storage
from .user.github import click_user_github
from .web_routes import (
    common,
    discover,
    fallback,
    new,
    update,
    user as web_user,
)

log = logging.getLogger(__name__)

CONTEXT_SETTINGS = {"help_option_names": ["-h", "--help"]}


class ErrorOnlyAccessLogger(AccessLogger):
    def log(self, request, response, time):
        # Only log if the status was not successful
        if not (200 <= response.status < 400):
            super().log(request, response, time)


def set_death_signal():
    PR_SET_PDEATHSIG = 1

    libc = ctypes.CDLL(find_library("c"))
    libc.prctl(PR_SET_PDEATHSIG, signal.SIGTERM, 0, 0, 0)


async def _run_tusd(host, tusd_port, web_port, base_path):
    command = (
        f"tusd",
        f"--host",
        f"{host}",
        f"--port",
        f"{tusd_port}",
        f"--hooks-http",
        f"http://127.0.0.1:{web_port}/new-package/tusd-internal",
        f"--hooks-enabled-events",
        f"pre-create,post-create,post-finish",
        f"--base-path",
        f"{base_path}",
    )
    tusd_proc = await asyncio.create_subprocess_exec(
        command[0],
        *command[1:],
        stdout=asyncio.subprocess.DEVNULL,
        stderr=asyncio.subprocess.DEVNULL,
        preexec_fn=set_death_signal,
    )
    await tusd_proc.wait()


@click_additional_options
def click_logging():
    logging.basicConfig(
        format="%(asctime)s %(levelname)-8s %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO
    )


@click.command(context_settings=CONTEXT_SETTINGS)
@click_logging  # Should always be on top, as it initializes the logging
@click_sentry
@click.option(
    "--bind", help="The IP to bind the server to", multiple=True, default=["::1", "127.0.0.1"], show_default=True
)
@click.option("--web-port", help="Port of the web server.", default=80, show_default=True, metavar="PORT")
@click.option("--tusd-port", help="Port of the tus server.", default=1080, show_default=True, metavar="PORT")
@common.click_reload_secret
@click_cleanup_graceperiod
@click_storage
@click_content_save
@click_user_session
@click_user_github
@click.option("--validate", help="Only validate BaNaNaS files and exit.", is_flag=True)
def main(bind, web_port, tusd_port, validate):
    """
    Start the BaNaNaS API.

    Every option can also be set via an environment variable prefixed with
    BANANAS_API_; for example:

    BANANAS_API_RELOAD_SECRET="test" python -m bananas_api
    """

    if validate:
        return

    webapp = web.Application()
    webapp.add_routes(common.routes)
    webapp.add_routes(discover.routes)
    webapp.add_routes(new.routes)
    webapp.add_routes(update.routes)
    webapp.add_routes(web_user.routes)

    register_webroutes(webapp)

    # Always make sure "fallback" comes last. It has a catch-all rule.
    webapp.add_routes(fallback.routes)

    # Start tusd as part of the application
    loop = asyncio.get_event_loop()
    for host in bind:
        if ":" in host:
            host = f"[{host}]"
        loop.create_task(_run_tusd(host, tusd_port, web_port, "/new-package/tus/"))

    # Start aiohttp server
    web.run_app(webapp, host=bind, port=web_port, access_log_class=ErrorOnlyAccessLogger)


if __name__ == "__main__":
    main(auto_envvar_prefix="BANANAS_API")
